Data Bundles v0.1 Schema (1/2)

susheel opened the pull request:
Initial commit based on #221.

Advanced expansions i.e. `expand=shallow` & `expand=deep` will have to wait until OAPI3 (for anyOf support)

[Optional] Data Bundle Access URL Method added in #245

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +194,71 @@ definitions:
+          A string representing a version, some systems may use checksum, a
+          RFC3339 timestamp, or incrementing version number. For systems that do not
+          support versioning please use your update timestamp as your version.
+      mime-type:
```
I don't think  `mime-type` makes sense on bundles -- aren't they always just a collection of other objects? (The referenced objects can of course be typed, but that's different.)

susheel replied:
If you are able to access the data in a bundle via an `access_url` then mime-type may be important. It also allows future extensions for container formats that can be accessed as a bundle and or individually.

susheel replied:
Removed Bundle `mime-type` from this PR. Will place it in a different PR for discussion.

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +194,71 @@ definitions:
+        description: >-
           At least one checksum must be provided.
           The Data Bundle checksum is computed over all the checksums of the
           Data Objects that bundle contains.
```
Can we be more explicit about how to compute the checksum?
* Is the idea simple checksum string concatenation? E.g. if Bundle B1 contains Objects O1 and O2, with checksums "cs111" and "cs222", then do I compute the checksum of bundle B1 by running my hash function on "cs111cs222"?
* what order should I use in computing my checksum? E.g. if Bundle B2 contains the same two objects, but happens to return them in a different order in the `contents` array, does it have a different checksum? 
* does the checksum include the `name` field from `BundleObjects`? If not, are we okay with two bundles that return differently named objects having the same checksum, as long as the objects have the same contents?

susheel replied:
Description updated. The algorithm is based on a data bundle checksum is based on a concatenated alphabetically (hex-code) sorted checksum of data objects

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +194,71 @@ definitions:
           Data Objects that bundle contains.
+        items:
+          $ref: '#/definitions/Checksum'
+      access_methods:
```
Do we need `access_methods` for Bundles? Given that the payload of a bundle is the `contents` array, which is returned in the response to this method, what further access is needed?

susheel replied:
Yes, I think so. If the entire List of Data Objects in a Bundle existed in a FTP/S3/... directory, we would want to provide a `access url` to the whole directory. It would also help requesting access to the whole directory via a single `access_id`.

If such a directory does not exist, then the DRS provider simply provides an empty list of `access_methods`.

ddietterich replied:
I don't understand @susheel. You have stored the bundle in a file. But instead of producing that file contents as the response to this request, you want the client to make another request that will produce the Bundle response? So then you are not producing the contents array below?

susheel replied:
@ddietterich No, @dglazer's question is more about should Bundles have a`ccess_methods`. I believe the specification should not restrict this.
The `access_url` of a Bundle provides a way to access the Bundle `contents` byte array in a single call. Consider a folder example:
ftp://server.com/data/A
A
├── DO1
└── DO2
The Data Objects `DO1` & `DO2` will all have access_urls:
DO1.*.access_url: ftp://server.com/data/A/DO1
DO2.*.access_url: ftp://server.com/data/A/DO2

A client would hand these `access_url`s to ftp-client for download.

If I were to create a Bundle A with it's `contents` represented by `DO1` and `DO2`, then the question is should I be able to provide an access_url for the Bundle. My argument is that the specification should not restrict this use case.
A.access_url: ftp://server.com/data/A

A client would hand this `access_url` to ftp-client for download.

The service provider must be certain that directory structure provided by the `access_url` conforms to the structure described by the Bundle `contents`

dglazer replied:
See https://github.com/ga4gh/data-repository-service-schemas/pull/244#issuecomment-480517411

susheel replied:
Removed `access_methods` to be discussed in a different PR.

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -396,3 +405,15 @@ definitions:
+    type: object
+    properties:
+      name:
+        type: string
```
please add a description (since it's not obvious that this string represents the name that should be used when materializing the object, and that the name here overrides any name that might be associated with the object itself)

dglazer replied:
also -- should we be explicit that this string is not allowed to contain slashes?

susheel replied:
Description added.

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -396,3 +405,15 @@ definitions:
+      name:
+        type: string
+      id:
+        type: string
```
please add a description, clarifying:
1. that this is the ID of a DRS object
2. whether it can also be the ID of a DRS bundle (and if so, how I know which is which)

susheel replied:
Description added. I have added `type` to the BundleObject to help the user make this distinction.

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +194,71 @@ definitions:
+          $ref: '#/definitions/BundleObjects'
+    required:
+      - id
+      - size
```
this object doesn't have a `size` field

susheel replied:
I missed this from the object. It should have a `size` computed from the cumulative sizes of each Data Object

dglazer replied:
Hmm -- that means all implementors would need to compute the total size for all objects in a bundle before they could expose it via DRS. I guess that's okay, given that we also require them to compute the checksum, but I'd like to hear from at least a couple people planning to implement Bundle support before imposing that mandate.

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +194,71 @@ definitions:
+      - id
+      - size
+      - created
+      - updated
```
this shouldn't be required (to match Object)

susheel replied:
We need to have either `created/updated` or `version` to let the user know if they are accessing the latest version or an older version of the bundle.

dglazer replied:
That feels like it's taking us down the slippery slope of being opinionated about versioning, which we were trying to avoid for now. I suggest leaving `updated` as optional, and then starting a separate conversation in a new issue/PR about whether we want to make it required for **both** `Object` and `Bundle`. (Unless there's some argument for them being different? I don't see one).

susheel replied:
Okay. Moving `updated` to optional

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +194,71 @@ definitions:
+      - size
+      - created
+      - updated
+      - version
```
this shouldn't be required (to match Object)

susheel replied:
We need to have either `created/updated` or `version` to let the user know if they are accessing the latest version or an older version of the bundle.

dglazer replied:
Same comment as on `updated` -- I suggest making it optional for now, and then separately discussing a change to both `Object` and `Bundle`.

susheel replied:
Okay, I agree

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +194,71 @@ definitions:
-        description: |-
+        description: >-
           An identifier, unique to this Data Bundle
       object_ids:
```
isn't `object_ids` redundant, now that we have `contents` ?

susheel replied:
Good catch. I forgot to remove it. Removed in 03c1b01ba49227f8f15737b8a282876e40946f58

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +181,87 @@ definitions:
+        type: string
+        description: >-
+          A string providing the mime-type of the Data Bundle.
+          For example, "application/json".
```
I suggest using the `example:` YAML syntax for consistency, and slightly friendlier documentation formatting

susheel replied:
Updated

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +181,85 @@ definitions:
+        type: string
+        format: int64
+        description: |-
+          The computed cumulated size of all Data Objects in bytes.
```
Is this recursive? In other words, if the Bundle contains Bundles that contain Objects, is `size` expected to be based on walking the whole tree? Either way, let's clarify the description. (And nit -- I think "computed cumulated" should be "cumulative" or "total".)

susheel replied:
@dglazer Recursive

@ddietterich Providing this would give the user a handle to the total size of the Bundle they will be downloading. If having `access_methods` are included in the Bundle specification then having `size` is a must.

susheel replied:
Clarified Bundle > size in description. size represents only the cumulative size of all objects within the Bundle's contents field (so depth 1)

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +181,85 @@ definitions:
+        description: >-
+          At least one checksum must be provided.
+          The Data Bundle checksum is computed over a sorted concatenation of all the 
+          checksums of the Data Objects that bundle contains. The list of Data
```
Similar comment as on `size` -- is this recursive, meaning it's based on the full list of Objects all the way down? Or is it just based on the checksums of the top-level Objects and Bundles?

susheel replied:
@dglazer Just the top-level, i.e. contents

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +181,85 @@ definitions:
+        description: >-
+          At least one checksum must be provided.
+          The Data Bundle checksum is computed over a sorted concatenation of all the 
+          checksums of the Data Objects that bundle contains. The list of Data
```
Confirming that the checksum does *not* include the `name` field? If so, that means two Bundles that contain the same Objects, but laid out in completely different directory trees, would have the same checksum -- is that intentional? (Not sure what makes sense here.)

susheel replied:
@dglazer Yes, two Bundles with the same list of objects (in any order) should have the same checksum.

This would allow the user to lay out the same physical (byte) Bundles in different (named) ways based on their specific usecase.

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -396,3 +406,33 @@ definitions:
       license:
         type: object
         description: License information for the exposed API
+  BundleObjects:
```
I suggest `BundleObject` (singular), since it's only one thing.

susheel replied:
Updated.

dglazer commented:
For those following along -- there are several small refinements being discussed and resolved in line-by-line comments. A few of those comments (in particular the comment threads on `mime-type` and `access_methods`, and the related new method in #245) feel like they're part of a higher-level discussion of how a bundle is expected to be fetched.

* My original vision was a Bundle was simply a possibly-nested list of Objects, meant to be fetched iteratively using our existing mechanisms. 
* I could also easily imagine someone storing a logical collection of files inside a single Object, possibly compressed, and meant to be fetched monolithically using our existing mechanisms and then unpacked by the client.
* We *could* also support a hybrid, where a single Bundle can be fetched either iteratively or monolithically. I don't know if any of our drivers need that, but if so, I'd suggest a simpler implementation, where the Bundle has a new optional field (e.g `bundle_object_id`) that allows callers who don't want to iterate to use the existing methods to fetch it one go. That way our Bundle methods stay focused on the iterative case, and don't need to duplicate all the monolithic Object fetch cases.

ddietterich commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +181,85 @@ definitions:
+          For example, "application/json".
       checksums:
         type: array
+        description: >-
```
Why is this an array here, but a single value for Object? That seems inconsistent.
I think it should be a simple reference to the `Checksum` with the computation method as described below.

susheel replied:
@ddietterich A Bundle or Object can have an array of checksums calculated using any hashing algorithm, not just MD5. The Checksum object provides the value and type of algorithm used.

dglazer replied:
@ddietterich I think it's an array for `Object` also? (I agree with you they should match; I think they already do.)

ddietterich commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +181,85 @@ definitions:
+        description: A human readable description.
       aliases:
         type: array
+        description: A list of strings that can be used to identify this Data Bundle.
```
Probably need to incorporate the changes to align with our discussion in the meeting today.

dglazer replied:
And to align with our further discussion in the meeting **today**.

dglazer replied:
Or, since this PR doesn't actually change the text of the description, you could just remove the syntax change to the YAML, and then this edit won't conflict with the one in #246 -- you call which is easier.

susheel replied:
I changed the `aliases` description to align with #246

ddietterich commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -396,3 +406,33 @@ definitions:
+      id:
+        type: string
+        description: >-
+          This represents the DRS identifier that may point to a Data Object
```
Is DRS identifier well-defined somewhere? (Easily could be. I'm still finding my way around here...) I assume this is intended to be a complete DRS URI - not just the id component. Is that correct?

susheel replied:
Yes, it should be the complete DRS URI. I will clarify the description.

I think the DRS ID URI must be defined somewhere in the documentation.

susheel commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +181,85 @@ definitions:
+        type: string
         description: |-
-          The list of Data Objects that this Data Bundle contains.
+          A string that can be optionally used to name a Data Bundle.
```
@ddietterich It should be implied by the missing `required` tag. No harm in reiterating this in the attribute description. This is mainly to make sure users don't use the name as the Bundle or Object's primary identifier.

I think the naming scheme will be implementation-specific.

susheel commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +181,85 @@ definitions:
         type: string
         format: date-time
-        description: |-
+        description: >-
```
@ddietterich Yes Bundle creation and update time, NOT of its contents.

susheel commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +181,85 @@ definitions:
-          A string representing a version, some systems may use checksum, a RFC3339
-          timestamp, or incrementing version number. For systems that do not support
-          versioning please use your update timestamp as your version.
+        description: >-
```
> If I make a new version of the directory, does it get a new create time? In which case, the create and update time are the same?

Yes.

> Why would you ask non-versioning systems to use the update timestamp for a version? That seems arbitrary. I'd prefer to use something that matches what I plan to use for versions if/when we do them.

It was a remnant from the older Bundle description. I have updated the description.

dglazer commented:
@ddietterich -- that's my prior also.
Susheel, are you picturing a use case equivalent to `zip -r filename.zip folder` as an optional alternate way to fetch the directory tree?

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +181,84 @@ definitions:
+          At least one checksum must be provided.
+          The Data Bundle checksum is computed over a sorted concatenation of all the 
+          checksums within the top-level 'contents' of the Bundle (not recursive). The list of Data
+          Object checksums are sorted alphabetically (hex-code) before concatenation and a further
```
As written, the checksum only includes the checksums of the top-level Objects -- is that right? I would have expected it to also include the checksums of any top-level Bundles. Maybe just replace "Data Object checksums" with "Object or Bundle checksums"?

susheel replied:
Tx. Description updated.

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -396,3 +405,33 @@ definitions:
+      id:
+        type: string
+        description: >-
+          This represents the DRS identifier URI that may point to a Data
```
Is this a URI? I expected this to be an `id` string that can be passed to `GET /objects/{object_id}` or `GET /bundles/{bundle_id}`.

susheel replied:
BundleObject updated with a `drs_uri` that represents a list of possible URIs to obtain the data object or bundle.

briandoconnor commented:
@susheel: Idea from conversation today, move the access_methods from this PR to a new PR to simplify this one.

He's also going to clarify documentation for URIs for objects rather than ID on its own.

susheel commented:
@dglazer Cleaned up the PR based on today's call:
- Added `drs_uri` field to BundleObject (represents a list of possible DRS URI paths to obtain the object or bundle
- Removed `mime-type` and `access_methods` for a separate discussion
- Clarified Bundle > `size` and Bundle > `checksum` descriptions.

dglazer commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -194,57 +181,77 @@ definitions:
+        description: >-
+          At least one checksum must be provided.
+          The Data Bundle checksum is computed over a sorted concatenation of all 
+          the checksums within the top-level 'contents' of the Bundle (not recursive).
```
@susheel -- confirming that this checksum does **not** include the `name` field from the `contents` array? (I"m not sure which makes more sense -- as written it means that two bundles that contain the same objects, but with different names, will end up having the same checksum as each other.)

susheel replied:
Clarified description to confirm names are NOT included in the checksum computation. Yes, two bundles containing the same objects should resolve to have the same checksum. The names assigned by the user should not affect the checksum, just how the individual objects are materialised onto disk.

dglazer commented:
@susheel , this is looking very close. Two unresolved comments in the code:
- confirm that Bundle `checksums` do **not** take `name` fields into account
- fix a merge conflict in the description of `aliases` (between this PR and #246)

ddietterich commented on openapi/data_repository_service.swagger.yaml:
```diff
@@ -396,3 +403,40 @@ definitions:
+          This represents the DRS identifier that may point to a Data
+          Object or another Data Bundle that will need to be recursed.
+      drs_uri:
+        type: array
```
I think the array needs an
```items:
    type: string
```

susheel replied:
Thanks @ddietterich. Resolved.
